agent/hcp/telemetry/otel_sink.go

```diff
@@ -0,0 +1,189 @@
+	Logger hclog.Logger
+}
+
+type OTELSink struct {
+	spaceReplacer *strings.Replacer
+	logger        hclog.Logger
+
+	meterProvider *otelsdk.MeterProvider
+	meter         *otelmetric.Meter
+
+	gaugeStore *gaugeStore
+
+	gaugeInstruments     map[string]*instrument.Float64ObservableGauge
+	counterInstruments   map[string]*instrument.Float64Counter
+	histogramInstruments map[string]*instrument.Float64Histogram
+
+	mutex sync.Mutex
+}
+
+func NewOTELReader(client client.MetricsClient, endpoint string, exportInterval time.Duration) otelsdk.Reader {
+	exporter := NewOTELExporter(client, endpoint)
+	return otelsdk.NewPeriodicReader(exporter, otelsdk.WithInterval(exportInterval))
+}
+
+func NewOTELSink(opts *OTELSinkOpts) (*OTELSink, error) {
+	if opts.Logger == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Logger")
+	}
+
+	if opts.Reader == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Reader")
+	}
+
+	// Setup OTEL Metrics SDK to aggregate, convert and export metrics periodically.
+	res := resource.NewSchemaless()
+	meterProvider := otelsdk.NewMeterProvider(otelsdk.WithResource(res), otelsdk.WithReader(opts.Reader))
+	meter := meterProvider.Meter("github.com/hashicorp/consul/agent/hcp/telemetry")
+
+	gs := &gaugeStore{
+		store: make(map[string]*gaugeValue, 0),
+		mutex: sync.Mutex{},
+	}
+
+	return &OTELSink{
+		spaceReplacer:        strings.NewReplacer(" ", "_"),
+		logger:               opts.Logger.Named("otel_sink"),
+		meterProvider:        meterProvider,
+		meter:                &meter,
+		mutex:                sync.Mutex{},
+		gaugeStore:           gs,
+		gaugeInstruments:     make(map[string]*instrument.Float64ObservableGauge, 0),
+		counterInstruments:   make(map[string]*instrument.Float64Counter, 0),
+		histogramInstruments: make(map[string]*instrument.Float64Histogram, 0),
+	}, nil
+}
+
+// SetGauge emits a Consul gauge metric.
+func (o *OTELSink) SetGauge(key []string, val float32) {
+	o.SetGaugeWithLabels(key, val, nil)
+}
+
+// AddSample emits a Consul histogram metric.
+func (o *OTELSink) AddSample(key []string, val float32) {
+	o.AddSampleWithLabels(key, val, nil)
+}
+
+// IncrCounter emits a Consul counter metric.
+func (o *OTELSink) IncrCounter(key []string, val float32) {
+	o.IncrCounterWithLabels(key, val, nil)
+}
+
+// AddSampleWithLabels emits a Consul gauge metric that gets
+// registed by an OpenTelemetry Histogram instrument.
+func (o *OTELSink) SetGaugeWithLabels(key []string, val float32, labels []gometrics.Label) {
+	k := o.flattenKey(key)
+
+	// Set value in global Gauge store.
+	o.gaugeStore.Store(k, float64(val), toAttributes(labels))
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	// If instrument does not exist, create it and register callback to emit last value in global Gauge store.
+	if _, ok := o.gaugeInstruments[k]; !ok {
+		inst, err := (*o.meter).Float64ObservableGauge(k, instrument.WithFloat64Callback(o.gaugeStore.gaugeCallback(k)))
```

Comment:
Can we document that this call back is really just registering the gauge for the first time? After that it should be in the store.